### ISSUE TITLE
Disable automatic passing down of props

### ISSUE THREAD

**Issue by timsuchanek** [styled-components/styled-components#135]
## Version

1.0.5
## Steps to reproduce

``` js
    import { Link } from 'react-router'

    const ListElement = styled(Link)`

      ${props => props.active && activeListElement}
    `
```
## Expected Behavior

Work without warnings
## Actual Behavior

<img width="589" alt="screen shot 2016-10-24 at 1 55 05 pm" src="https://cloud.githubusercontent.com/assets/1094804/19644693/813e2eb2-99f1-11e6-9bd0-517ad44e3177.png">

Is it possible to disable the automatic pass down of props so we don't have that warning anymore?

Thanks!

**Comment by vdanchenkov** [comment]
Do you mean some props selectively? I'd suggest:
```jsx
const ListElement = styled(({ active, ...rest }) => <Link {...rest} />)`
```

**Comment by timsuchanek** [comment]
Yes indeed. Nice solution, thanks!

**Comment by timsuchanek** [comment]
As we are using typescript and it doesn't yet support the spread operator in objects, thats my solution right now:
```js
export function ExcludeProps(SubComponent: any, filter: string[] = []) {
  return (props) => {
    const keys = Object.keys(props).filter(key => !filter.includes(key))
    const picked = lodash.pick(props, keys)

    return <SubComponent {...picked} />
  }
}
```
@mxstbr Maybe we should write an official statement into the docs as other people could encounter a similiar problem with the new react version?

**Comment by mxstbr** [comment]
Hmm, that's an issue with out `validAttrs.js` util – it seems to let that one pass through, even though it's invalid.

**Comment by vdanchenkov** [comment]
@mxstbr This issue is about styling 3-rd party components. I believe that you want to pass all the props unfiltered in this case.

**Comment by mxstbr** [comment]
Ah yeah, true – I think this is actually expected behaviour here, you're right! I'll close this for now since @vdanchenkov's solution works fine and we can't really do anything from our side to fix this. Cheers!

**Comment by mxstbr** [comment]
Ah yeah, true – I think this is actually expected behaviour here, you're right! I'll close this for now since @vdanchenkov's solution works fine and we can't really do anything from our side to fix this. Cheers!

**Comment by alex-cory** [comment]
I did mine like this.
```
import { withRouter } from 'react-router-dom'
import styled from 'styled-components'


export const Link = withRouter(({ to, history, onClick = null, ...props }) =>
	<Base {...props} onClick={e => onClick ? onClick(e) : history.push(to)} />
)

const Base = styled.a`
	cursor: pointer;
	text-decoration: none;
`
```

**Comment by joetidee** [comment]
What about if you are extending a component? In the example below, I want `completed` to be accessed by `StyledItemTextarea` but not `Textarea`.
```
import Textarea from 'react-textarea-autosize';

const TextareaAutosizeSC = styled(Textarea)`
    ...
`;

const StyledItemTextarea = TextareaAutosizeSC.extend`
    color: ${({ completed }) => completed ? '#ccc' : '#fff'};
`;
```

**Comment by nicgirault** [comment]
```javascript
const ListElement = styled(({ active, ...rest }) => <Link {...rest} />)`
```

works fine (although the syntax  highlighting is broken on vscode). However it creates an anonymous component that makes it harder to navigate in the DOM and read a stacktrace. Naming the component is much more verbose and does not solve the ident in the DOM so I would support an option to intercept props.

With a very naïve approach I would like to be able to write something like:

```javascript
const ListElement = styled(Link, {intercept: ['active']})`
```

**Comment by janosh** [comment]
Any comments from maintainers regarding @nicgirault's proposal? Something along those lines would be nice to have!

**Comment by alansouzati** [comment]
I'm facing this problem too. I don't think `const ListElement = styled(({ active, ...rest }) => <Link {...rest} />)` is bullet proof solution. Here is where it is failing for me:

**1) Callers can pass `as` prop and now the component suddenly breaks. for example:**

```
<ListElement as="span" />
```

One would expect that it would render a `Link` as a span tag, but it gets rid of the Link altogether. Don't get me wrong, this is the expected behavior in styled-components, but it makes it difficult customize things. I had to introduce a `tagName` prop just for that. And I keep getting the question why you need "tagName" since styled-components has "as" prop?

**2) Hard, if not impossible to extend styles.**

I have a Box component with some styles. This Box has a `color` prop that I need to exclude (otherwise it will bleed). Now I want a Flex component that augments this Box with flexbox capabilities. This `Flex` has a `direction` prop that I need to exclude. For example:

```
const Box = styled(({ color, ...rest}) => <div {...rest} />)`
  color: ${props => props.color};
`;

const Flex = styled(Box)`
  flex-direction: ${props => props.direction};
`;
```

Now if you inspect the DOM Box color does not bleed, but Flex direction does. I'm failing to identify how to properly extend these styles.

**3) Having to deal with forwardRef and increasing the React DOM Tree.**

Since I'm adding `({ active, ...rest }) => <Link {...rest} />` forward refs don't work anymore. Meaning `<ListElement ref={React.createRef()} />`the ref will be the React element instead of the actual DOM, forcing me to use a deprecated `findDOMNode` API. Or wire the forwardRef which increases the React Depth Tree ([which brings performance issues in React dev env](https://github.com/facebook/react/issues/13739))

I feel that everything would be much easier if styled-component allowed an API to exclude props to bleed to the DOM:

```
const Box = styled("div", { excludeProps: ['color'] })`
  color: ${props => props.color};
`;

const Flex = styled(Box, { excludeProps: ['direction'] })`
  flex-direction: ${props => props.direction};
`;
````

**Comment by dmackerman** [comment]
Can't agree with @alansouzati more.

**Comment by hejiaji** [comment]
No solution yet?

**Comment by JamesBliss** [comment]
I see this is closed but doesn't really feel like a good solid solution to stopping prop bleed is available.

I might have missed something though, can I get an update on whether this is still be looked at?

**Comment by tomspeak** [comment]
Agreed, I am facing this issue every other day and the workarounds are starting to add up.

**Comment by JamesBliss** [comment]
@tomspeak I've found these:

A v5 Roadmap #1934 which links to issue #2093 but nothing still seems defined, decided or clear.

**Comment by JamesBliss** [comment]
@tomspeak and anyone else... here is an example of how it could be bypassed as an issue. Although this solution would be very hard to get a team to understand and stick too. [codesandbox example!](https://codesandbox.io/s/0pjrmn6nw0)

**Comment by rtymchyk** [comment]
This is very annoying when trying to make components with AMP :)

**Comment by Alappin** [comment]
A temporary clean work around.

```

import styled from 'styled-components';
import omit from 'lodash/omit';

const withFilter(Tag, filter = []) => {
	return props => <Tag {...omit(props, filter)} />;
};

const Tag = ...;

const StyledTag = styled(withFilter(Tag, ['color'])`
  color: ${props => props.color}
` 
```
Skipping the filter array will just gracefully work. I guess the ideal solution in the future would just allow the second parameter to be used for filtering as others have already suggested in the thread above.

**Comment by nkov** [comment]
Another workaround is to use `data-*` attributes, for which the DOM doesn't throw warnings.

**Comment by nathggns** [comment]
I don't think this issue should be closed. I'm very much still running into this issue, and it seems unfortunate that the proper way to avoid this problem is to add a wrapper component.

**Comment by Zerquix18** [comment]
+1 ??

**Comment by jdfm** [comment]
@Alappin  `excludeProps` makes me think that the prop will not be available, so, when writing out the CSS my expectation is that it's not there for our use. How about `captureProps` instead?

**Comment by rtymchyk** [comment]
> A temporary clean work around.

This is good but but `omit` is being dropped for performance reasons.

Here's a variation:

```javascript
import React from 'react'
import PropTypes from 'prop-types'
import pickBy from 'lodash/pickBy'

export default function omitProps(Component, propsToOmit) {
  function isAllowedProp(_, prop) {
    return !propsToOmit.includes(prop)
  }

  function WithoutOmittedProps({ children, ...rest }) {
    const withOmittedProps = pickBy(rest, isAllowedProp)
    return <Component {...withOmittedProps}>{children}</Component>
  }

  WithoutOmittedProps.propTypes = {
    children: PropTypes.node,
  }

  WithoutOmittedProps.displayName = `WithoutOmittedProps(${Component.displayName || Component.name || Component})`

  return WithoutOmittedProps
}
```

**Comment by lcswillems** [comment]
I have just written an RFC for filtering forwarded props: https://github.com/styled-components/styled-components/issues/2878

Would you like a solution like this?

**TL;DR** It consists of adding a second parameter for filtering props. The current solution:

```js
const StyledDiv = styled(({height, width, ...props}) => <div {...props} />)`
    width: ${props => props.width}px;
    height: ${props => props.height}px;
`
```

is turned into:

```js
const StyledDiv = styled.div(`
    width: ${props => props.width}px;
    height: ${props => props.height}px;
`, ({width, height, ...props}) => props);
```

This is easy to write and to understand, and props filtering (which is not the important part of the definition) is sent at the end of the definition.

**Comment by random1911** [comment]
Hey guys.
What about made new API similar to Emotion - https://emotion.sh/docs/styled#customizing-prop-forwarding - that shouldForwardProp really helpfull with damn <Link> styling

**Comment by SimonSomlai** [comment]
Here's a simple typescript solution that worked for me;

```javascript
interface LinkProps {
  to: string;
}

interface SidebarLinkProps extends LinkProps {
  active: boolean;
}

const SidebarLink = styled(({ active, ...rest }: SidebarLinkProps) => (
  <RouteLink {...rest} />
))`
  color: red;
  ${({ active }) => active && `color: green`}
`;
```

**Comment by webcoderkz** [comment]
@SimonSomlai how do you type `<RouteLink />`'s children? I'm just getting an error: `ype '{ children: string; }' has no properties in common with type`

**Comment by anks333** [comment]
You can use transient props, Checkout this docs

https://styled-components.com/docs/api#transient-props

**Comment by jonatassales** [comment]
Created a solution based on what @anks333 did:

`
import { StyledConfig } from 'styled-components'

export default function omitForwardedProps(props: string[]): StyledConfig {
  return {
    shouldForwardProp: (prop, defPropValFN) => !props.includes(prop) && defPropValFN(prop)
  }
}
`

`
import styled from 'styled-components'
import Typography, { TypographyProps } from '@material-ui/core/Typography'

import { omitForwardedProps } from '@infra/styled-components/utils'
interface BreadcrumbItemProps extends TypographyProps {
  highlighted?: boolean
}

export const BreadcrumbItem = styled(Typography).withConfig(omitForwardedProps(['highlighted']))<BreadcrumbItemProps>`
  color: ${({ highlighted, theme }) => (highlighted ? theme.colors.primary : theme.colors.black)};
  font-size: ${({ theme }) => theme.fontSizes.headline6};
  white-space: nowrap;
`

`
